plugins/ltac/extratactics.mlg

```diff
@@ -1119,3 +1119,28 @@ let tclOPTIMIZE_HEAP =
 TACTIC EXTEND optimize_heap
 | [ "optimize_heap" ] -> { tclOPTIMIZE_HEAP }
 END
+
+(** tactic analogous to "Strategy" *)
+
+{
+
+let set_strategy lvl ql =
+  let glob_key r =
+    match r with
+    | GlobRef.ConstRef sp -> ConstKey sp
+    | GlobRef.VarRef id -> VarKey id
+    | _ -> user_err Pp.(str
+        "cannot set an inductive type or a constructor as transparent") in
+  List.iter (fun q -> Global.set_strategy (glob_key q) lvl) ql
```

Comment:
Global modification of the environment should not be allowed from tactic code.

Reply:
Is the solution to move this part of the global state to proof state?  (I tried figuring out how to do that but couldn't make heads or tails of what's in proof state and what's not.)

Reply:
@ejgallego Ping

Reply:
Proof state is what is on the `engine/proofview` monad, we had some discussion about it in the Coq Call.

In order to add the state you would indeed augment the state monad there, but IMHO that's a risky proposition and that could have some performance impact; I dunno, the thing doesn't seem to be designed as to be extensible. Relevant code in `proofview_monad.ml`:
```
module P = struct

  type s = proofview * Environ.env

```
see also `logic_monad.ml`

Reply:
The strategy stuff is already in the env, why do you say to change that part of the code?

Reply:
To provide a bit more of context, what we are trying to do for now is to have the proof state not to depend on `Global`, you can see that the state may not be synchronized properly at some points, for example w.r.t. environmnets; I need to sync with @ppedrot to see what the steps are [I have some patches sitting on my HD], but in short; we'd like to finish pushing up `Global` up before we move towards making the proof state easier to tweak [in a sense the more things we put inside the larger the risk for problems later when we attempt the move]

Reply:
> The strategy stuff is already in the env, why do you say to change that part of the code?

Oh sorry, indeed you had mentioned that in the call; so it may be doable without modifying the state indeed.

Reply:
>Oh sorry, indeed you had mentioned that in the call; so it may be doable without modifying the state indeed.

The problem is that @JasonGross wants this to work
~~~coq
strategy bla.
foo_tac. (* use the new strategy *)
~~~
but the env in the monad gets reset after every `.` using the global env.

Reply:
Indeed, thanks for the reminder @SkySkimmer ; holidays have done a bit of damage to my brain. I guess that's the reason I mentioned the cleanups, I have a patch in my HD that should help with this problem, as it tries to improve how the envs are handled, I guess I'll try to finish it soon.